File: Utils/src/main/java/com/colin/library/android/utils/PathUtil.java
//package com.colin.library.android.utils;
//
//import android.content.Context;
//import android.os.Build;
//import android.os.Environment;
//
//import androidx.annotation.NonNull;
//import androidx.annotation.Nullable;
//
//import com.colin.library.android.utils.data.Constants;
//import com.colin.library.android.helper.UtilHelper;
//
//import java.io.File;
//
///**
// * 作者： ColinLu
// * 时间： 2022-01-15 22:42
// * <p>
// * 描述： Android 路径辅助类
// */
//public final class PathUtil {
//    private PathUtil() {
//        throw new UnsupportedOperationException("don't instantiate");
//    }
//
//    /*判断是否存在外部存储卡*/
//    public static boolean storageState() {
//        return Environment.MEDIA_MOUNTED.equals(Environment.getExternalStorageState());
//    }
//
//    /*外部存储是否有写权限*/
//    public static boolean canWrite() {
//        return storageState() && Environment.getExternalStorageDirectory().canWrite();
//    }
//
//    /*  /system */
//    public static File getRootSystem() {
//        return Environment.getRootDirectory();
//    }
//
//    /*  /storage*/
//    public static File getRootStorage() {
//        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
//            return Environment.getStorageDirectory();
//        }
//        return null;
//    }
//
//
//    /*  /data */
//    public static File getRootData() {
//        return Environment.getDataDirectory();
//    }
//
//    /*  /data/cache*/
//    public static File getDownloadCache() {
//        return Environment.getDownloadCacheDirectory();
//    }
//
//
//    /*/data/data/package*/
//    @NonNull
//    public static File getInternalApp() {
//        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
//            return new File(UtilHelper.getInstance().getContext().getApplicationInfo().dataDir);
//        }
//        return UtilHelper.getInstance().getContext().getDataDir();
//    }
//
//    /*/data/user/0/package*/
//    public static File getUserData() {
//        return getUserData(UtilHelper.getInstance().getContext());
//    }
//
//    /*/data/user/0/package*/
//    public static File getUserData(@NonNull Context context) {
//        return context.getCodeCacheDir();
//    }
//
//    /*/data/user/0/package/cache*/
//    public static File getUserCache() {
//        return getUserCache(UtilHelper.getInstance().getContext());
//    }
//
//    /*/data/user/0/package/cache*/
//    public static File getUserCache(@NonNull Context context) {
//        return context.getCacheDir();
//    }
//
//    /*/data/user/0/package/code_cache*/
//    public static File getUserCodeCache() {
//        return getUserCodeCache(UtilHelper.getInstance().getContext());
//    }
//
//    /*/data/user/0/package/code_cache*/
//    public static File getUserCodeCache(@NonNull Context context) {
//        return context.getCodeCacheDir();
//    }
//
//    /*/data/user/0/package/files*/
//    public static File getUserFiles() {
//        return getUserFiles(UtilHelper.getInstance().getContext());
//    }
//
//    /*/data/user/0/package/files*/
//    public static File getUserFiles(@NonNull Context context) {
//        return context.getFilesDir();
//    }
//
//    /*  storage/emulated/0*/
//    public static File getExternalStorage() {
//        return Environment.getExternalStorageDirectory();
//    }
//
//    /*  storage/emulated/0/type */
//    public static File getExternalFile(@NonNull String type) {
//        return Environment.getExternalStoragePublicDirectory(type);
//    }
//
//    /**
//     * /storage/emulated/0/Android/data/package/files    type == null
//     * /storage/emulated/0/Android/data/package/files/type
//     *
//     * @param context
//     * @param type
//     * @return
//     */
//    public static File getExternalFile(@NonNull Context context, @Nullable String type) {
//        return context.getExternalFilesDir(type);
//    }
//
//    /*/storage/emulated/0/Android/data/package/cache*/
//    public static File getExternalCache() {
//        return getExternalCache(UtilHelper.getInstance().getContext());
//    }
//
//    /*/storage/emulated/0/Android/data/package/cache*/
//    public static File getExternalCache(@NonNull Context context) {
//        return context.getExternalCacheDir();
//    }
//
//
//    @NonNull
//    public static String getInternalAppPath() {
//        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
//            return UtilHelper.getInstance().getContext().getApplicationInfo().dataDir;
//        }
//        return UtilHelper.getInstance().getContext().getDataDir().getAbsolutePath();
//    }
//
//    /*清除缓存 操作 /data/data/package/cache */
//    @NonNull
//    public static File getInternalCache() {
//        return UtilHelper.getInstance().getContext().getCacheDir();
//    }
//
//    /*/data/data/package/shared_prefs*/
//    public static String getInternalAppSpPath() {
//        return getInternalAppPath() + Constants.FILE_SEP + "shared_prefs";
//    }
//
//    /*/data/data/package/code_cache*/
//    public static String getInternalAppCodeCache() {
//        return getInternalAppPath() + Constants.FILE_SEP + "code_cache";
//    }
//
//    @Nullable
//    public static String getPath(@Nullable final File file) {
//        return file == null ? null : file.getPath();
//    }
//
//    @Nullable
//    public static String getAbsolutePath(@Nullable final File file) {
//        return file == null ? null : file.getAbsolutePath();
//    }
//}

File: Utils/src/main/java/com/colin/library/android/utils/PathUtil.kt
package com.colin.library.android.utils

import android.annotation.SuppressLint
import android.content.Context
import android.os.Build
import android.os.Environment
import com.colin.library.android.utils.helper.UtilHelper
import java.io.File

/**
 * Author:ColinLu
 * E-mail:945919945@qq.com
 * Create:2025-06-29 20:13
 *
 * Des   :PathUtil
 */
object PathUtil {
    private const val DB_FILE = "databases"
    private const val SP_FILE = "shared_prefs"
    private const val CODE_FILE = "code_cache"

    /*判断是否存在外部存储卡*/
    fun hasSDCard() = Environment.MEDIA_MOUNTED == Environment.getExternalStorageState()

    /*外部存储是否有写权限*/
    fun canWrite(): Boolean {
        return hasSDCard() && Environment.getExternalStorageDirectory().canWrite()
    }

    /*  /system */
    fun getRootSystem() = Environment.getRootDirectory()

    /*  /data */
    fun getRootData() = Environment.getDataDirectory()

    /*/data/user/0/package*/
    @SuppressLint("ObsoleteSdkInt")
    fun getInternalData(context: Context = UtilHelper.getApplication()): File? {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) context.dataDir else File(context.applicationInfo.dataDir)
    }

    /*/data/user/0/package/files*/
    fun getInternalFiles(context: Context = UtilHelper.getApplication()) = context.filesDir

    /*/data/data/package/cache */
    fun getInternalCache(context: Context = UtilHelper.getApplication()) = context.cacheDir

    /*/data/user/0/package/code_cache*/
    fun getInternalCode(context: Context = UtilHelper.getApplication()) =
        File(getInternalPath(context, CODE_FILE))

    /*/data/user/0/package/databases*/
    fun getInternalDatabase(context: Context = UtilHelper.getApplication()) =
        File(getInternalPath(context, DB_FILE))

    /*/data/user/0/package/shared_prefs*/
    fun getInternalSp(context: Context = UtilHelper.getApplication()) =
        File(getInternalPath(context, SP_FILE))


    /*  storage/emulated/0*/
    fun getExternalStorage(): File? {
        return Environment.getExternalStoragePublicDirectory(null)
    }

    /**
     * /storage/emulated/0/Android/data/package/files/type
     * type==null-->>/storage/emulated/0/Android/data/package/files
     * type==xxx -->>/storage/emulated/0/Android/data/package/files/xxx
     * @param context
     * @param type
     * @return
     */
    fun getAppExternalFile(context: Context = UtilHelper.getApplication(), type: String?) =
        context.getExternalFilesDir(type)

    /*/storage/emulated/0/Android/data/package/cache*/
    fun getExternalCache(context: Context = UtilHelper.getApplication()) = context.externalCacheDir


    /*  /data/cache*/
    fun getDownloadCache() = Environment.getDownloadCacheDirectory()

    /**
     * /storage/emulated/0/type
     * type==null-->>/storage/emulated/0
     * type==xxx -->>/storage/emulated/0/xxx
     * @param type #
     * @return
     */
    fun getStorageExternalFile(type: String): File? {
        return Environment.getExternalStoragePublicDirectory(type)
    }

    @SuppressLint("ObsoleteSdkInt")
    fun getInternalPath(context: Context = UtilHelper.getApplication()): String {
        return if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
            context.applicationInfo.dataDir
        } else context.dataDir.absolutePath
    }

    fun getInternalPath(context: Context, name: String) =
        "${getInternalPath(context)}${File.separator}$name"

    fun getPath(file: File?): String? {
        return file?.path
    }

    fun getAbsolutePath(file: File?): String? {
        return file?.absolutePath
    }

}